pkg/controller/aws/network/securitygroup/controller.go

```diff
@@ -0,0 +1,227 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	runtimev1alpha1 "github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane-runtime/pkg/resource"
+	v1alpha1 "github.com/crossplaneio/crossplane/aws/apis/network/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/clients/aws/ec2"
+	"github.com/crossplaneio/crossplane/pkg/controller/aws/utils"
+)
+
+const (
+	errUnexpectedObject = "The managed resource is not an SecurityGroup resource"
+)
+
+// Controller is the controller for SecurityGroup objects
+type Controller struct{}
+
+// SetupWithManager creates a new Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func (c *Controller) SetupWithManager(mgr ctrl.Manager) error {
+	r := resource.NewManagedReconciler(mgr,
+		resource.ManagedKind(v1alpha1.SecurityGroupGroupVersionKind),
+		resource.WithExternalConnecter(&connector{client: mgr.GetClient(), newClientFn: ec2.NewSecurityGroupClient, awsConfigFn: utils.RetrieveAwsConfigFromProvider}),
+		resource.WithManagedConnectionPublishers())
+	name := strings.ToLower(fmt.Sprintf("%s.%s", v1alpha1.SecurityGroupKindAPIVersion, v1alpha1.Group))
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.SecurityGroup{}).
+		Complete(r)
+}
+
+type connector struct {
+	client      client.Client
+	newClientFn func(*aws.Config) (ec2.SecurityGroupClient, error)
+	awsConfigFn func(context.Context, client.Reader, *corev1.ObjectReference) (*aws.Config, error)
+}
+
+func (conn *connector) Connect(ctx context.Context, mgd resource.Managed) (resource.ExternalClient, error) {
+	cr, ok := mgd.(*v1alpha1.SecurityGroup)
+	if !ok {
+		return nil, errors.New(errUnexpectedObject)
+	}
+
+	awsconfig, err := conn.awsConfigFn(ctx, conn.client, cr.Spec.ProviderReference)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := conn.newClientFn(awsconfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create a new SecurityGroupClient")
+	}
+
+	return &external{c}, nil
+}
+
+type external struct {
+	client ec2.SecurityGroupClient
+}
+
+func (e *external) Observe(ctx context.Context, mgd resource.Managed) (resource.ExternalObservation, error) {
+	cr, ok := mgd.(*v1alpha1.SecurityGroup)
+	if !ok {
+		return resource.ExternalObservation{}, errors.New(errUnexpectedObject)
+	}
+
+	// To find out whether a SecurityGroup exist:
+	// - the object's ExternalState should have securityGroupId populated
+	// - a SecurityGroup with the given securityGroupId should exist
+	if cr.Status.SecurityGroupID == "" {
+		return resource.ExternalObservation{
+			ResourceExists: false,
+		}, nil
+	}
+
+	req := e.client.DescribeSecurityGroupsRequest(&awsec2.DescribeSecurityGroupsInput{
+		GroupIds: []string{cr.Status.SecurityGroupID},
+	})
+	req.SetContext(ctx)
+
+	response, err := req.Send()
+
+	if err != nil {
+		if ec2.IsSecurityGroupNotFoundErr(err) {
```

Comment:
These functions look like they handle the case in which `err == nil`, so we could probably avoid nesting this clause.

Reply:
good catch!